add tutorial "Setup iota.js for Development" (#720)
iota-wiki/iota-wiki: content/tutorials/single-page-tutorials/setup_iota-js_ for_development.md

# New Tutorial - How to setup iota.js for development

> > In this tutorial you will learn how to create a new project, generate a new address and receive testnet funds from the faucet.

Reviewer: I think we can use an info admonition here

```suggestion
:::info

In this tutorial you will learn how to create a new project, generate a new address and receive testnet funds from the faucet.

:::
```

## Prerequisites

- Development Environment and Libraries
  - Installed [Node.js](https://nodejs.org/en/)

***

## Setup

### Prepare development environment

Create a new folder for the tutorial in a location of your choice and navigate to it:

```bash
mkdir iota-js-setup
cd iota-js-setup
```

Run the Node.js initializer to configure the project:

```bash
npm init --yes
```

Install the iota.js library and other helper dependencies:

> > **NOTE:**  Installation might take a while

Reviewer: ```suggestion
:::note

Installation might take a while

:::
```

```bash
npm install @iota/iota.js@next
npm install bip39 dotenv
```

Now create a new file `.env` and add the content below:

```javascript
MNEMONIC = "<You_will_create_your_own_mnemonic_seed_phrase_in_the_next_step_and_paste_it_in_here>"
```

> Enter your desired account name as well as a secure Stronghold password. Your new mnemonic seed phrase will be created in the next step and you will paste it in here afterwards.

Reviewer: Where should I add a stronghold password and account name? :trollface: 

```suggestion
Your new mnemonic seed phrase will be created in the next step and you will paste it in here afterwards.
```

***

## Create mnemonic script

Create a new file `create-mnemonic.js` and add the following code:

```javascript
// Libraries
const bip39 = require('bip39')
 
// For better readability, some console output will be printed in a different color
const consoleColor = '\x1b[36m%s\x1b[0m';
async function run() {
	// A 256-bits entropy leads to a 24 word mnemonic seed phrase
	const mnemonic = bip39.generateMnemonic(256);
	console.log(consoleColor, 'Copy your mnemonic seed phrase and paste it into the .env file:');
	console.log(mnemonic, '\n');
}
run();
```

Run the script `create-mnemonic.js` and check the console output for your new mnemonic seed phrase:
```console
node create-mnemonic.js
```
Now you can copy the seed phrase and paste it into the previously created `.env` file.

***

Your `package.json` file should look something like this:

```yaml
{
  "name": "iota-js-setup",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "dependencies": {
    "@iota/iota.js": "^1.9.0-stardust.25",
    "bip39": "^3.0.4",
    "dotenv": "^16.0.1"
  },
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "Apache-2.0"
}
```

> ***

Reviewer: Shouldn't this be moved to the beginning after installing the devs?

Author: It's just an example how it should look like after the installation part.

Reviewer: Sure I got that. I was just surprised reading it after already having done the installation some time ago 🤔

Reviewer 2: I must say I agree with @Dr-Electron , this seems awfully out of place here as we're in the generate mnemonic section by now.

Author: Fixed!

## Create an address

Create a new file `create-address.js` and add the following code:

> ```javascript!

Reviewer: ```suggestion
```javascript
```

const API_ENDPOINT = "https://api.testnet.shimmer.network";

async function run() {
  const client = new SingleNodeClient(API_ENDPOINT);

  const info = await client.info();

  console.log("Base");

  // Generate the seed from the Mnemonic
  const baseSeed = Ed25519Seed.fromMnemonic(process.env.MNEMONIC);
  console.log("\tSeed", Converter.bytesToHex(baseSeed.toBytes()));

  // Generate the next addresses for your account.
  console.log();
  console.log("Generated Addresses...");
  const addressGeneratorAccountState = {
    accountIndex: 0,
    addressIndex: 0,
    isInternal: false,
  };
  const path = generateBip44Address(addressGeneratorAccountState);
  const addressSeed = baseSeed.generateSeedFromPath(new Bip32Path(path));
  const addressKeyPair = addressSeed.keyPair();
  const indexEd25519Address = new Ed25519Address(addressKeyPair.publicKey);
  const indexPublicKeyAddress = indexEd25519Address.toAddress();

  console.log(
    "\tAddress (Bech32)",
    Bech32Helper.toBech32(
      ED25519_ADDRESS_TYPE,
      indexPublicKeyAddress,
      info.protocol.bech32Hrp
    )
  );
}

run()
  .then(() => console.log("Done"))
  .catch((err) => console.error(err));

```

> If you run it, it will get your mnemonic from the `.env`file, creates a seed from it and genrates a new address. Just run the following command and you will see your testnet shimmer address in your console!

Reviewer: ```suggestion
If you run it, it will get your mnemonic from the `.env`file, create a seed from it and generate a new address. Just run the following command and you will see your testnet shimmer address in your console!
```

> ```bash!

Reviewer: ```suggestion
```bash
```

> # New Tutorial - How to setup iota.js for development

Reviewer 2: ```suggestion
# New Tutorial - How to Set Up Iota.JS for Development
```
Do we need the "new tutorial" in the headline?

:::info

In this tutorial you will learn how to create a new project, generate a new address and receive testnet funds from the faucet.

:::

## Prerequisites

> - Development Environment and Libraries
>   - Installed [Node.js](https://nodejs.org/en/)

Reviewer 2: ```suggestion
### Development Environment and Libraries
  - [Node.js](https://nodejs.org/en/)
```
do we require a specific or minimum version?

***

> ## Setup

Reviewer 2: ```suggestion
## Set Up
```
You set up your setup (I know, it's confusing). But basically, if it's instructions it's Set Up, if it's describing the required "tools" it's setup

> ### Prepare development environment

Reviewer 2: ```suggestion
### Prepare Your Development Environment
```

> Create a new folder for the tutorial in a location of your choice and navigate to it:

Reviewer 2: ```suggestion
1. Create a new folder for the tutorial in a location of your choice and navigate to it:
```

```bash
mkdir iota-js-setup
cd iota-js-setup
```

> Run the Node.js initializer to configure the project:

Reviewer 2: ```suggestion
2. Run the Node.js initializer to configure the project:
```

```bash
npm init --yes
```

> Install the iota.js library and other helper dependencies:

Reviewer 2: ```suggestion
3. Install the iota.js library and other helper dependencies:
```

:::note

> Installation might take a while

Reviewer 2: ```suggestion
The installation process might take a while
```

:::

> ```bash
> npm install @iota/iota.js@next
> npm install bip39 dotenv
> ```

Reviewer 2: ````suggestion

```bash
npm install @iota/iota.js@next
npm install bip39 dotenv
```
````

Reviewer 2: we may want to consider putting the code before the note.

> Now create a new file `.env` and add the content below:

Reviewer 2: ```suggestion
4. Create a new `.env` file and add the content below:
```

> ```javascript

Reviewer 2: Just leaving a comment in case this wasn't intentional.

MNEMONIC = "<You_will_create_your_own_mnemonic_seed_phrase_in_the_next_step_and_paste_it_in_here>"
```

> Your new mnemonic seed phrase will be created in the next step and you will paste it in here afterwards.

Reviewer 2: Wouldn't it make more sense to give the create mnemonic instructions before?

***

> ## Create mnemonic script

Reviewer 2: ```suggestion
## Create Mnemonic Script
```

> Create a new file `create-mnemonic.js` and add the following code:

Reviewer 2: ```suggestion
1. Create a new file `create-mnemonic.js` and add the following code:
```

```javascript
// Libraries
const bip39 = require('bip39')
 
// For better readability, some console output will be printed in a different color
const consoleColor = '\x1b[36m%s\x1b[0m';
async function run() {
	// A 256-bits entropy leads to a 24 word mnemonic seed phrase
	const mnemonic = bip39.generateMnemonic(256);
	console.log(consoleColor, 'Copy your mnemonic seed phrase and paste it into the .env file:');
	console.log(mnemonic, '\n');
}
run();
```

> Run the script `create-mnemonic.js` and check the console output for your new mnemonic seed phrase:

Reviewer 2: ```suggestion
2. Run the script `create-mnemonic.js` and check the console output for your new mnemonic seed phrase:

```

```console
node create-mnemonic.js
```

> Now you can copy the seed phrase and paste it into the previously created `.env` file.

Reviewer 2: ```suggestion

3. Copy the seed phrase and paste it into the previously created `.env` file.
```

***

Your `package.json` file should look something like this:

> ```yaml

Reviewer 2: ```suggestion
```json
```

{
  "name": "iota-js-setup",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "dependencies": {
    "@iota/iota.js": "^1.9.0-stardust.25",
    "bip39": "^3.0.4",
    "dotenv": "^16.0.1"
  },
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "Apache-2.0"
}
```
***

> ## Create an address

Reviewer 2: ```suggestion
## Create an Address
```

> Create a new file `create-address.js` and add the following code:

Reviewer 2: ```suggestion
Create a new file called `create-address.js` and add the following code:
```

const API_ENDPOINT = "https://api.testnet.shimmer.network";

async function run() {
  const client = new SingleNodeClient(API_ENDPOINT);

  const info = await client.info();

  console.log("Base");

  // Generate the seed from the Mnemonic
  const baseSeed = Ed25519Seed.fromMnemonic(process.env.MNEMONIC);
  console.log("\tSeed", Converter.bytesToHex(baseSeed.toBytes()));

  // Generate the next addresses for your account.
  console.log();
  console.log("Generated Addresses...");
  const addressGeneratorAccountState = {
    accountIndex: 0,
    addressIndex: 0,
    isInternal: false,
  };
  const path = generateBip44Address(addressGeneratorAccountState);
  const addressSeed = baseSeed.generateSeedFromPath(new Bip32Path(path));
  const addressKeyPair = addressSeed.keyPair();
  const indexEd25519Address = new Ed25519Address(addressKeyPair.publicKey);
  const indexPublicKeyAddress = indexEd25519Address.toAddress();

  console.log(
    "\tAddress (Bech32)",
    Bech32Helper.toBech32(
      ED25519_ADDRESS_TYPE,
      indexPublicKeyAddress,
      info.protocol.bech32Hrp
    )
  );
}

run()
  .then(() => console.log("Done"))
  .catch((err) => console.error(err));

```

> If you run it, it will get your mnemonic from the `.env`file, create a seed from it and generate a new address. Just run the following command and you will see your testnet shimmer address in your console!

Reviewer 2: ```suggestion
If you run it, it will:

1. Get your mnemonic from the `.env`file.
2. Create a seed from it.
3. Generate a new address. 

Just run the following command, and you will see your testnet shimmer address in your console!
```

```bash
node create-address.js
```

> ## Request tokens via faucet website

Reviewer 2: ```suggestion
## Request Tokens via the Faucet Website
```

> Now you can go to the [Shimmer Testnet Faucet Website](https://faucet.testnet.shimmer.network/), paste in your newly created address and request funds.

Reviewer 2: ```suggestion
Now, you can go to the [Shimmer Testnet Faucet Website](https://faucet.testnet.shimmer.network/), paste your newly created address, and request funds.
```

:::note

The 'Request' button will become clickable as soon as a valid Shimmer address is recognized.

:::
![Shimmer Testnet Faucet](./images/shimmer-testnet-faucet.png)

***

> Congratulations, you are now ready to explore other exciting features of the Shimmer network, such as native tokens and NFT's.

Reviewer 2: This seems like a nice place to add CTAs to the other tutorials once they're published.
